File: vscode-extension/src/EventListenerManager.ts
import * as vscode from 'vscode';
import {ActionType, LogFormatter} from './Type';
import {Logger} from './Logger';
import {EditorStateManager} from './EditorStateManager';
import {FileUtils} from './FileUtils';
import {WindowStateManager} from './WindowStateManager';

/**
 * Pending TAB open events
 */
interface PendingTabEvent {
    uri: vscode.Uri;
    tab: vscode.Tab;
    timeout: NodeJS.Timeout;
}

/**
 * Event listener manager
 * Unified management of various VSCode editor event listeners
 */
export class EventListenerManager {
    private disposables: vscode.Disposable[] = [];
    private logger: Logger;
    private editorStateManager: EditorStateManager;
    private windowStateManager: WindowStateManager;

    // Pending TAB event queue, used to distinguish foreground/background opening
    private pendingTabEvents: Map<string, PendingTabEvent> = new Map();

    // Delay confirmation time (milliseconds)
    private readonly CONFIRMATION_DELAY = 150;

    constructor(
        logger: Logger,
        editorStateManager: EditorStateManager,
        windowStateManager: WindowStateManager
    ) {
        this.logger = logger;
        this.editorStateManager = editorStateManager;
        this.windowStateManager = windowStateManager;
    }


    /**
     * Set up editor listeners
     */
    setupEditorListeners() {
        this.logger.info('Setting up editor listeners');
        // Listen for active editor changes
        this.disposables.push(
            vscode.window.onDidChangeActiveTextEditor((editor) => {
                if (!editor) {
                    return;
                }
                if (!FileUtils.isRegularFileUri(editor.document.uri)) {
                    return;
                }

                const filePath = editor.document.uri.fsPath;

                // Check if there is a corresponding pending TAB event
                const pendingEvent = this.pendingTabEvents.get(filePath);
                if (pendingEvent) {
                    // Cancel pending TAB event because this is foreground opening
                    clearTimeout(pendingEvent.timeout);
                    this.pendingTabEvents.delete(filePath);
                    this.logger.info(`Canceled pending TAB event, confirmed as foreground opening: ${filePath}`);
                }

                this.logger.info(`Event - File foreground opened: ${filePath}`);
                const state = this.editorStateManager.createEditorState(
                    editor, ActionType.OPEN, this.windowStateManager.isWindowActive()
                );
                this.logger.info(`Preparing to send foreground open message: ${state.filePath}`);
                this.editorStateManager.updateState(state);
            })
        );

        // Listen for TAB closure
        this.disposables.push(
            vscode.window.tabGroups.onDidChangeTabs((event) => {
                // Handle newly opened TABs
                event.opened.forEach((tab, index) => {
                    // Detect if tab type is regular file, ignore other types
                    if (!FileUtils.isRegularFileTab(tab)) {
                        this.logger.info(`Opened TAB ${index}: Non-regular file type, ignored`);
                        return
                    }

                    const uri = (tab.input as vscode.TabInputText).uri;
                    const filePath = uri.fsPath;

                    this.logger.info(`Event - File TAB opened: ${filePath}`);

                    // Create timeout handler for delayed confirmation
                    const timeout = setTimeout(() => {
                        // Confirmed as background opening after timeout
                        this.pendingTabEvents.delete(filePath);
                        this.handleBackgroundOpen(uri);
                    }, this.CONFIRMATION_DELAY);

                    // Add to pending queue
                    this.pendingTabEvents.set(filePath, {
                        uri,
                        tab,
                        timeout
                    });

                    this.logger.info(`TAB event added to pending queue: ${filePath}`);
                });

                event.closed.forEach((tab, index) => {
                    // Detect if tab type is regular file, ignore other types
                    if (!FileUtils.isRegularFileTab(tab)) {
                        this.logger.info(`Closed TAB ${index}: Non-regular file type, ignored`);
                        return;
                    }
                    const uri = (tab.input as vscode.TabInputText).uri;
                    this.logger.info(`Event - File closed: ${uri.fsPath}`);
                    const filePath = uri.fsPath;

                    // Check if file is still open in other TABs
                    const isStillOpen = FileUtils.isFileOpenInOtherTabs(filePath);
                    if (isStillOpen) {
                        this.logger.info(`File still open in other TABs, skipping close message: ${filePath}`);
                        return;
                    }

                    this.logger.info(`File completely closed, sending close message: ${filePath}`);
                    const state = this.editorStateManager.createCloseState(
                        filePath,
                        this.windowStateManager.isWindowActive()
                    );
                    this.logger.info(`Preparing to send close message: ${state.filePath}`);
                    this.editorStateManager.updateState(state)
                });
            })
        )

        // Listen for cursor position and selection changes
        this.disposables.push(
            vscode.window.onDidChangeTextEditorSelection((event) => {
                if (!FileUtils.isRegularFileUri(event.textEditor.document.uri)) {
                    return;
                }

                const hasSelection = !event.textEditor.selection.isEmpty;
                const selection = event.textEditor.selection;
                const cursorPosition = selection.active;
                const filePath = event.textEditor.document.uri.fsPath;

                let logMessage = `Event - Selection changed: ${filePath}, ${LogFormatter.cursorLog(cursorPosition.line, cursorPosition.character)}`;

                if (hasSelection) {
                    logMessage += `，${LogFormatter.selectionLog(selection.start.line, selection.start.character, selection.end.line, selection.end.character)}`;
                } else {
                    logMessage += `，${LogFormatter.selectionLog()}`;
                }
                this.logger.info(logMessage);

                if (event.textEditor === FileUtils.getCurrentActiveEditor()) {
                    const state = this.editorStateManager.createEditorState(
                        event.textEditor, ActionType.NAVIGATE, this.windowStateManager.isWindowActive()
                    );
                    this.logger.info(`Preparing to send navigation message: ${state.action} ${state.filePath}, ${state.getSelectionLog()}, ${state.getSelectionLog()}`);
                    this.editorStateManager.debouncedUpdateState(state);
                }
            })
        );

        this.logger.info('Editor listeners setup completed');
    }


    /**
     * Handle background file opening
     */
    private handleBackgroundOpen(uri: vscode.Uri) {
        const filePath = uri.fsPath;

        this.logger.info(`Event - File background opened: ${filePath}`);

        // Send file open event
        const openState = this.editorStateManager.createEditorStateFromPath(
            filePath, ActionType.OPEN, this.windowStateManager.isWindowActive()
        );
        this.logger.info(`Preparing to send background open message: ${openState.filePath}`);
        this.editorStateManager.updateState(openState);


        const navigateState = this.editorStateManager.getCurrentActiveEditorState(this.windowStateManager.isWindowActive());
        if (navigateState) {
            this.logger.info(`Preparing to send current active editor navigation message: ${navigateState.filePath}`);
            this.editorStateManager.debouncedUpdateState(navigateState);
        }
    }


    /**
     * Clean up resources
     */
    dispose() {
        // Clean up all pending TAB events
        this.pendingTabEvents.forEach((pendingEvent) => {
            clearTimeout(pendingEvent.timeout);
        });
        this.pendingTabEvents.clear();

        this.disposables.forEach(d => d.dispose());
        this.disposables = [];
    }
}

File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/EventListenerManager.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerEvent
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.util.messages.MessageBusConnection

/**
 * Event listener manager
 * Unified management of various editor event listeners, converting events to standard operation tasks
 */
class EventListenerManager(
    private val project: Project,
    private val editorStateManager: EditorStateManager,
    private val windowStateManager: WindowStateManager,
    private val fileUtils: FileUtils
) {
    private val log: Logger = Logger.getInstance(EventListenerManager::class.java)

    // Global unique cursor listener reference
    private var currentCaretListener: com.intellij.openapi.editor.event.CaretListener? = null

    // Global unique selection listener reference
    private var currentSelectionListener: com.intellij.openapi.editor.event.SelectionListener? = null
    private var currentEditor: Editor? = null
    private var messageBusConnection: MessageBusConnection? = null


    /**
     * Setup editor listeners
     */
    fun setupEditorListeners() {
        log.info("Setting up editor listeners")
        messageBusConnection = project.messageBus.connect();
        messageBusConnection?.subscribe(
            FileEditorManagerListener.FILE_EDITOR_MANAGER,
            object : FileEditorManagerListener {
                override fun fileOpened(source: FileEditorManager, file: VirtualFile) {
                    if (!fileUtils.isRegularFile(file)) {
                        log.info("Event - File opened: ${file.path} - Irregular file, ignored")
                        return
                    }
                    log.info("Event - File opened: ${file.path}")
                    val fileEditor = source.getSelectedEditor(file)
                    val editor = if (fileEditor is com.intellij.openapi.fileEditor.TextEditor) {
                        fileEditor.editor
                    } else {
                        null
                    }
                    editor?.let {
                        val state = editorStateManager.createEditorState(
                            it, file, ActionType.OPEN, windowStateManager.isWindowActive()
                        )
                        log.info("Preparing to send open message: $state")
                        editorStateManager.updateState(state)
                        setupCaretListener(it)
                        setupSelectionListener(it)
                        currentEditor = it
                    }
                }

                override fun fileClosed(source: FileEditorManager, file: VirtualFile) {
                    if (!fileUtils.isRegularFile(file)) {
                        log.info("Event - File closed: ${file.path} - Irregular file, ignored")
                        return
                    }
                    log.info("Event - File closed: ${file.path}")

                    // Check if file is still open in other editors
                    val isStillOpen = fileUtils.isFileOpenInOtherTabs(file)
                    if (isStillOpen) {
                        log.info("File is still open in other editors, skipping close message: ${file.path}")
                        return
                    }

                    // Create close state and send to queue (no dependency on editor object)
                    val state = editorStateManager.createCloseState(file.path, windowStateManager.isWindowActive())
                    log.info("Preparing to send close message: $state")
                    editorStateManager.updateState(state)
                }

                override fun selectionChanged(event: FileEditorManagerEvent) {
                    if (event.newFile != null) {
                        if (!fileUtils.isRegularFile(event.newFile!!)) {
                            log.info("Event - File changed: ${event.newFile!!.path} - Irregular file, ignored")
                            return
                        }
                        log.info("Event - File changed: ${event.newFile!!.path}")
                        val fileEditor = event.newEditor
                        val editor = if (fileEditor is com.intellij.openapi.fileEditor.TextEditor) {
                            fileEditor.editor
                        } else {
                            null
                        }
                        editor?.let {
                            val state = editorStateManager.createEditorState(
                                it, event.newFile!!, ActionType.OPEN, windowStateManager.isWindowActive()
                            )
                            log.info("Preparing to send open message: ${state.filePath}，${state.getCursorLog()}，${state.getSelectionLog()}")
                            editorStateManager.updateState(state)
                            setupCaretListener(it)
                            setupSelectionListener(it)
                            currentEditor = it
                        }
                    }
                }
            }
        )
        log.info("Editor listeners setup completed")
    }


    /**
     * Unified handling of editor cursor and selection events
     * @param editor Editor instance
     * @param eventType Event type ("cursor changed" or "selection changed")
     */
    private fun handleEditorPositionOrSelectionEvent(editor: Editor, eventType: String) {
        log.info("Event-$eventType")

        // Dynamically get current actual file
        val currentFile = editor.virtualFile
        if (currentFile == null) {
            log.warn("Event-$eventType: Unable to get current file, skipping processing")
            return
        }
        if (!fileUtils.isRegularFile(currentFile)) {
            log.info("Event-$eventType: ${currentFile.path} - Irregular file, ignored")
            return
        }

        // Get cursor position information
        val caretModel = editor.caretModel
        val logicalPosition = caretModel.logicalPosition
        var logMessage = "Event-$eventType: ${currentFile.path}，${LogFormatter.cursorLog(logicalPosition.line, logicalPosition.column)}"

        // Add selection information (regardless of event type)
        val selectionModel = editor.selectionModel
        val hasSelection = selectionModel.hasSelection()

        if (hasSelection) {
            val startPosition = editor.offsetToLogicalPosition(selectionModel.selectionStart)
            val endPosition = editor.offsetToLogicalPosition(selectionModel.selectionEnd)
            logMessage += "，${LogFormatter.selectionLog(startPosition.line, startPosition.column, endPosition.line, endPosition.column)}"
        } else {
            logMessage += "，${LogFormatter.selectionLog(null, null, null, null)}"
        }

        log.info(logMessage)

        val state = editorStateManager.createEditorState(
            editor, currentFile, ActionType.NAVIGATE, windowStateManager.isWindowActive()
        )
        log.info("Preparing to send navigation message: ${state.filePath}，${state.getCursorLog()}，${state.getSelectionLog()}")
        editorStateManager.debouncedUpdateState(state)
    }


    /**
     * Setup cursor listener
     * Globally unique, destroys previous listener when setting up new one
     */
    private fun setupCaretListener(editor: Editor) {
        log.info("Starting to setup cursor listener")

        // Destroy previous cursor listener
        destroyCurrentCaretListener()

        // Create new cursor listener
        val newCaretListener = object : com.intellij.openapi.editor.event.CaretListener {
            override fun caretPositionChanged(event: com.intellij.openapi.editor.event.CaretEvent) {
                handleEditorPositionOrSelectionEvent(event.editor, "cursor changed")
            }
        }

        // Add new listener
        editor.caretModel.addCaretListener(newCaretListener)

        // Save reference for subsequent management
        currentCaretListener = newCaretListener

        log.info("Cursor listener setup completed")
    }

    /**
     * Setup selection listener
     * Globally unique, destroys previous listener when setting up new one
     */
    private fun setupSelectionListener(editor: Editor) {
        log.info("Starting to setup selection listener")

        // Destroy previous selection listener
        destroyCurrentSelectionListener()

        // Create new selection listener
        val newSelectionListener = object : com.intellij.openapi.editor.event.SelectionListener {
            override fun selectionChanged(event: com.intellij.openapi.editor.event.SelectionEvent) {
                handleEditorPositionOrSelectionEvent(event.editor, "selection changed")
            }
        }

        // Add new listener
        editor.selectionModel.addSelectionListener(newSelectionListener)

        // Save reference for subsequent management
        currentSelectionListener = newSelectionListener

        log.info("Selection listener setup completed")
    }

    /**
     * Destroy current cursor listener
     */
    private fun destroyCurrentCaretListener() {
        if (currentCaretListener != null && currentEditor != null) {
            log.info("Destroying previous cursor listener")
            try {
                currentEditor!!.caretModel.removeCaretListener(currentCaretListener!!)
                log.info("Cursor listener destroyed successfully")
            } catch (e: Exception) {
                log.warn("Exception occurred when destroying cursor listener: ${e.message}")
            }
            currentCaretListener = null
        }
    }

    /**
     * Destroy current selection listener
     */
    private fun destroyCurrentSelectionListener() {
        if (currentSelectionListener != null && currentEditor != null) {
            log.info("Destroying previous selection listener")
            try {
                currentEditor!!.selectionModel.removeSelectionListener(currentSelectionListener!!)
                log.info("Selection listener destroyed successfully")
            } catch (e: Exception) {
                log.warn("Exception occurred when destroying selection listener: ${e.message}")
            }
            currentSelectionListener = null
        }
    }


    /**
     * Clean up resources
     */
    fun dispose() {
        log.info("Starting to clean up EventListenerManager resources")
        messageBusConnection?.disconnect()
        messageBusConnection?.dispose()
        destroyCurrentCaretListener()
        destroyCurrentSelectionListener()
        currentEditor = null
        log.info("EventListenerManager resource cleanup completed")
    }
}
